File: 0016-3sum-closest/0016-3sum-closest.go
func threeSumClosest(nums []int, target int) int {
    if len(nums) < 3 {return -1}
    
    sort.Ints(nums)
    smallestDiff := 1 << 31
    
    for i := 0; i < len(nums) - 2; i++ {
        left, right := i + 1, len(nums) - 1
        
        for left < right {
            diff := target - nums[i] - nums[left] - nums[right]
            if diff == 0 {return target}
            
            if abs(diff) < abs(smallestDiff) || (abs(diff) == abs(smallestDiff) && diff > smallestDiff) {
                smallestDiff = diff
            }
            
            if diff > 0 {
                left++
            } else {
                right--
            }
        }
    }
    
    return target - smallestDiff
}

func abs(x int) int {
    if x > 0 {
        return x
    }
    
    return -x
}

File: 0016-3sum-closest/0016-3sum-closest.php
class Solution {

    /**
     * @param Integer[] $nums
     * @param Integer $target
     * @return Integer
     */
    function threeSumClosest($nums, $target) {
        $diff  = PHP_INT_MAX;
        $n = count($nums);
        sort($nums);

        for($i = 0;$i <= $n;$i++){
            $low = $i + 1;
            $high = $n - 1;
            while($low < $high) {
                $sum = $nums[$i] + $nums[$low] +$nums[$high];
                if(abs($target- $sum) < abs($diff)) {
                    $diff = $target- $sum;
                }
                if($sum < $target) {
                    $low++;
                } else {
                    $high--;
                }
            }
        }
        return $target - $diff;
    }
}